### PR TITLE
Loopify recursive checkForDuplicateSchemas call to avoid heap crash

### PR DESCRIPTION
**Pull request by daviwil**
This change addresses #285 which reports that running AutoRest with Modelerfour 4.13.351 on the SQL service spec causes a fatal heap allocation error:

```
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xa0e670 node::Abort() [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 2: 0xa0ea9c node::OnFatalError(char const*, char const*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 3: 0xb83afe v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 4: 0xb83e79 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 5: 0xd32305  [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 6: 0xd32996 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 7: 0xd41209 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 8: 0xd42045 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
 9: 0xd44b1c v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
10: 0xd0bd22 v8::internal::Handle<v8::internal::FixedArray> v8::internal::Factory::NewFixedArrayWithMap<v8::internal::FixedArray>(v8::internal::RootIndex, int, v8::internal::AllocationType) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
11: 0xf2c4b9 v8::internal::HashTable<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::New(v8::internal::Isolate*, int, v8::internal::AllocationType, v8::internal::MinimumCapacity) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
12: 0xf34c3e v8::internal::Dictionary<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::Add(v8::internal::Isolate*, v8::internal::Handle<v8::internal::NameDictionary>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyDetails, int*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
13: 0xf34db9 v8::internal::BaseNameDictionary<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::AddNoUpdateNextEnumerationIndex(v8::internal::Isolate*, v8::internal::Handle<v8::internal::NameDictionary>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyDetails, int*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
14: 0xf37313 v8::internal::BaseNameDictionary<v8::internal::NameDictionary, v8::internal::NameDictionaryShape>::Add(v8::internal::Isolate*, v8::internal::Handle<v8::internal::NameDictionary>, v8::internal::Handle<v8::internal::Name>, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyDetails, int*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
15: 0xefff72 v8::internal::LookupIterator::ApplyTransitionToDataProperty(v8::internal::Handle<v8::internal::JSReceiver>) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
16: 0xf314b1 v8::internal::Object::AddDataProperty(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::Maybe<v8::internal::ShouldThrow>, v8::internal::StoreOrigin) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
17: 0xedb110 v8::internal::JSObject::DefineOwnPropertyIgnoreAttributes(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::Maybe<v8::internal::ShouldThrow>, v8::internal::JSObject::AccessorInfoHandling) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
18: 0xedb349 v8::internal::JSObject::DefineOwnPropertyIgnoreAttributes(v8::internal::LookupIterator*, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::internal::JSObject::AccessorInfoHandling) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
19: 0xe21ed0 v8::internal::JsonParser<unsigned short>::BuildJsonObject(v8::internal::JsonParser<unsigned short>::JsonContinuation const&, std::vector<v8::internal::JsonProperty, std::allocator<v8::internal::JsonProperty> > const&, v8::internal::Handle<v8::internal::Map>) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
20: 0xe22a83 v8::internal::JsonParser<unsigned short>::ParseJsonValue() [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
21: 0xe2393f v8::internal::JsonParser<unsigned short>::ParseJson() [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
22: 0xc36f05 v8::internal::Builtin_JsonParse(int, unsigned long*, v8::internal::Isolate*) [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
23: 0x13da5f9  [/opt/hostedtoolcache/node/13.14.0/x64/bin/node]
Process() cancelled due to failure 
Invoke: /home/vsts/work/1/s/eng/Generation.psm1:45
Line |
  45 |      Invoke $command
     |      ~~~~~~~~~~~~~~~
     | Command failed to execute: /home/vsts/work/1/s/node_modules/.bin/autorest-beta --require=/home/vsts/work/1/s/readme.md https://github.com/Azure/azure-rest-api-specs/blob/d9cf7c7ed3d674ebd482836e82b274014245ae67/specification/network/resource-manager/readme.md --namespace=network_resource_manager --output-folder=/home/vsts/work/1/s/samples/smoketests/network-resource-manager/network-resource-manager
```

The fix is to convert the recursive invocation of `checkForDuplicateSchemas` into a loop so that the call stack no longer preserves huge amounts of data when processing large service specs.  I do this by creating a nested function called `innerCheckForDuplicateSchemas` that is invoked in a `while` loop and runs until this function returns something other than `true` (which indicates a duplicate was found and the scan should be performed again).

### CODE REVIEW

File: modelerfour/quality-precheck/prechecker.ts [hunk 1]
```diff
@@ -185,62 +185,73 @@ export class QualityPreChecker {
...
   checkForDuplicateSchemas(): undefined {
     const errors = new Set<string>();
-    if (this.input.components && this.input.components.schemas) {
-      const dupedNames = items(this.input.components?.schemas).select(s => ({ key: s.key, value: this.resolve(s.value) })).groupBy(each => each.value.instance['x-ms-metadata']?.name, each => each);
-      for (const [name, schemas] of dupedNames.entries()) {
-        if (name && schemas.length > 1) {
 
+    // Returns true if scanning should be restarted
+    const innerCheckForDuplicateSchemas = (): any => {
+      if (this.input.components && this.input.components.schemas) {
```
Comment (daviwil) [inline modelerfour/quality-precheck/prechecker.ts:192]: These are almost entirely indentation changes.

File: modelerfour/quality-precheck/prechecker.ts [hunk 2]
```diff
@@ -185,62 +185,73 @@ export class QualityPreChecker {
...
+                  primaryMetadata.apiVersions = [...new Set<string>([...primaryMetadata.apiVersions || [], ...secondaryMetadata.apiVersions || []])]
+                  primaryMetadata.filename = [...new Set<string>([...primaryMetadata.filename || [], ...secondaryMetadata.filename || []])]
+                  primaryMetadata.originalLocations = [...new Set<string>([...primaryMetadata.originalLocations || [], ...secondaryMetadata.originalLocations || []])]
+                  primaryMetadata['x-ms-secondary-file'] = !(!primaryMetadata['x-ms-secondary-file'] || !secondaryMetadata['x-ms-secondary-file'])
+                }
               }
+              this.session.verbose(`Schema ${name} has multiple identical declarations, reducing to just one - removing ${schemas[1].key} `, ['PreCheck', 'ReducingSchema']);
+
+              // Restart the scan now that the duplicate has been removed
+              return true;
```
Comment (daviwil) [inline modelerfour/quality-precheck/prechecker.ts:226]: This is a new change in this PR.

**Comment by pakrym** [comment]
Seems like this version requires deduplication flags enabled where it wasn't required before. For example in
https://github.com/Azure/azure-rest-api-specs/blob/d9cf7c7ed3d674ebd482836e82b274014245ae67/specification/web/resource-manager/readme.md and https://github.com/Azure/azure-rest-api-specs/blob/d9cf7c7ed3d674ebd482836e82b274014245ae67/specification/monitor/resource-manager/readme.md

**Comment by pakrym** [comment]
And errors are very strange too. There are multiple errors for the same schema with different set of properties.

```
ERROR (PreCheck/DuplicateSchema): Duplicate Schema named HttpRequestInfo -- properties.clientRequestId.$ref: undefined => "#/components/schemas/schemas:398",properties.clientIpAddress.$ref: undefined => "#/components/schemas/schemas:399",properties.method.$ref: undefined => "#/components/schemas/schemas:400",properties.uri.$ref: undefined => "#/components/schemas/schemas:401" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named HttpRequestInfo -- properties.clientIpAddress.$ref: undefined => "#/components/schemas/schemas:399",properties.method.$ref: undefined => "#/components/schemas/schemas:400",properties.uri.$ref: undefined => "#/components/schemas/schemas:401" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named HttpRequestInfo -- properties.method.$ref: undefined => "#/components/schemas/schemas:400",properties.uri.$ref: undefined => "#/components/schemas/schemas:401" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named HttpRequestInfo -- properties.uri.$ref: undefined => "#/components/schemas/schemas:401" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.claims.$ref: undefined => "#/components/schemas/schemas:403",properties.caller.$ref: undefined => "#/components/schemas/schemas:405",properties.description.$ref: undefined => "#/components/schemas/schemas:406",properties.id.$ref: undefined => "#/components/schemas/schemas:407",properties.eventDataId.$ref: undefined => "#/components/schemas/schemas:408",properties.correlationId.$ref: undefined => "#/components/schemas/schemas:409",properties.resourceGroupName.$ref: undefined => "#/components/schemas/schemas:411",properties.resourceId.$ref: undefined => "#/components/schemas/schemas:412",properties.operationId.$ref: undefined => "#/components/schemas/schemas:413",properties.properties.$ref: undefined => "#/components/schemas/schemas:414",properties.eventTimestamp.$ref: undefined => "#/components/schemas/schemas:416",properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.caller.$ref: undefined => "#/components/schemas/schemas:405",properties.description.$ref: undefined => "#/components/schemas/schemas:406",properties.id.$ref: undefined => "#/components/schemas/schemas:407",properties.eventDataId.$ref: undefined => "#/components/schemas/schemas:408",properties.correlationId.$ref: undefined => "#/components/schemas/schemas:409",properties.resourceGroupName.$ref: undefined => "#/components/schemas/schemas:411",properties.resourceId.$ref: undefined => "#/components/schemas/schemas:412",properties.operationId.$ref: undefined => "#/components/schemas/schemas:413",properties.properties.$ref: undefined => "#/components/schemas/schemas:414",properties.eventTimestamp.$ref: undefined => "#/components/schemas/schemas:416",properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.description.$ref: undefined => "#/components/schemas/schemas:406",properties.id.$ref: undefined => "#/components/schemas/schemas:407",properties.eventDataId.$ref: undefined => "#/components/schemas/schemas:408",properties.correlationId.$ref: undefined => "#/components/schemas/schemas:409",properties.resourceGroupName.$ref: undefined => "#/components/schemas/schemas:411",properties.resourceId.$ref: undefined => "#/components/schemas/schemas:412",properties.operationId.$ref: undefined => "#/components/schemas/schemas:413",properties.properties.$ref: undefined => "#/components/schemas/schemas:414",properties.eventTimestamp.$ref: undefined => "#/components/schemas/schemas:416",properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.id.$ref: undefined => "#/components/schemas/schemas:407",properties.eventDataId.$ref: undefined => "#/components/schemas/schemas:408",properties.correlationId.$ref: undefined => "#/components/schemas/schemas:409",properties.resourceGroupName.$ref: undefined => "#/components/schemas/schemas:411",properties.resourceId.$ref: undefined => "#/components/schemas/schemas:412",properties.operationId.$ref: undefined => "#/components/schemas/schemas:413",properties.properties.$ref: undefined => "#/components/schemas/schemas:414",properties.eventTimestamp.$ref: undefined => "#/components/schemas/schemas:416",properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.eventDataId.$ref: undefined => "#/components/schemas/schemas:408",properties.correlationId.$ref: undefined => "#/components/schemas/schemas:409",properties.resourceGroupName.$ref: undefined => "#/components/schemas/schemas:411",properties.resourceId.$ref: undefined => "#/components/schemas/schemas:412",properties.operationId.$ref: undefined => "#/components/schemas/schemas:413",properties.properties.$ref: undefined => "#/components/schemas/schemas:414",properties.eventTimestamp.$ref: undefined => "#/components/schemas/schemas:416",properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.correlationId.$ref: undefined => "#/components/schemas/schemas:409",properties.resourceGroupName.$ref: undefined => "#/components/schemas/schemas:411",properties.resourceId.$ref: undefined => "#/components/schemas/schemas:412",properties.operationId.$ref: undefined => "#/components/schemas/schemas:413",properties.properties.$ref: undefined => "#/components/schemas/schemas:414",properties.eventTimestamp.$ref: undefined => "#/components/schemas/schemas:416",properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.resourceGroupName.$ref: undefined => "#/components/schemas/schemas:411",properties.resourceId.$ref: undefined => "#/components/schemas/schemas:412",properties.operationId.$ref: undefined => "#/components/schemas/schemas:413",properties.properties.$ref: undefined => "#/components/schemas/schemas:414",properties.eventTimestamp.$ref: undefined => "#/components/schemas/schemas:416",properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.resourceId.$ref: undefined => "#/components/schemas/schemas:412",properties.operationId.$ref: undefined => "#/components/schemas/schemas:413",properties.properties.$ref: undefined => "#/components/schemas/schemas:414",properties.eventTimestamp.$ref: undefined => "#/components/schemas/schemas:416",properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.operationId.$ref: undefined => "#/components/schemas/schemas:413",properties.properties.$ref: undefined => "#/components/schemas/schemas:414",properties.eventTimestamp.$ref: undefined => "#/components/schemas/schemas:416",properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.properties.$ref: undefined => "#/components/schemas/schemas:414",properties.eventTimestamp.$ref: undefined => "#/components/schemas/schemas:416",properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.eventTimestamp.$ref: undefined => "#/components/schemas/schemas:416",properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.submissionTimestamp.$ref: undefined => "#/components/schemas/schemas:417",properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.subscriptionId.$ref: undefined => "#/components/schemas/schemas:418",properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventData -- properties.tenantId.$ref: undefined => "#/components/schemas/schemas:419" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventDataCollection -- properties.value.$ref: undefined => "#/components/schemas/schemas:421",properties.nextLink.$ref: undefined => "#/components/schemas/schemas:422" 

ERROR (PreCheck/DuplicateSchema): Duplicate Schema named EventDataCollection -- properties.nextLink.$ref: undefined => "#/components/schemas/schemas:422" 

```

**Review by joheredi — APPROVED** [review]
Looks good! Thanks for adding comments it makes it easier to follow

**Comment by daviwil** [comment]
Thanks @joheredi!  I'm pretty certain this resolves the issue without any more side effects.  Merging this now to get folks unblocked.

**Pull request merged**